website/blog/2023-10-24-sqlsmith-hero.md

```diff
@@ -0,0 +1,104 @@
+- **DDL**: `CREATE`, `ALTER`, `DROP TABLE` etc. to create and modify test tables.
+- **DML**: `INSERT`, `UPDATE`, `DELETE`, `MERGE` etc. to populate test data.
+- **Query**: `WITH`, `SELECT`, `JOIN`, `SubQuery`, `ORDER BY` etc. `WITH` and `SubQuery` can generate complex queries.
+- **Expression**: `Column`, `Literal`, `Scalar Function` and `Aggregate Function` etc. Expressions can be nested.
+
+By randomly generating AST components, it covers all possible SQL syntax. Controlled recursion also creates complex nested SQLs to find obscure bugs. But depth is capped to avoid unexecutable SQLs.
+
+## SQL Reducer
+
+Since generated SQLs can be very complex with the bug trigger buried, simplifying the original SQL makes bug isolation easier.
+
+The SQL Reducer iteratively removes AST components like `WITH`, `SubQuery`, `Expression` etc. If the simplified SQL still reproduces the bug, it's used. Else original SQL is retained. Ultimately, it outputs the smallest SQL to reproduce the bug by traversing all AST parts.
+
+## Runner 
+
+The SQL Runner loops generating test SQLs, executes them using Databend, and processes failures as:
+
+- Check for expected errors like syntax, semantics etc.
+- Verify if it's a known issue or unimplemented feature.
+- Call the SQL Reducer to generate the smallest reproducible SQL.
+- Log errors and simplified SQLs.
+
+The Runner is integrated into Databend's CI, executing on every release to log issues for further analysis.
+
+## Impact
+
+Thus far, SQLsmith has uncovered [50+ bugs](https://github.com/datafuselabs/databend/issues?q=is%3Aissue+label%3Afound-by-sqlsmith) in Databend after a month of running, including:
+  
+- Internal logic errors (17 bugs)
+- Invalid function/expression checks (12 bugs)  
+- Missing semantic checks (9 bugs)
+- Improper `unwrap`/`unreachable` handling (7 bugs) 
+- Parser failures (3 bugs)
+- Failed casts across types (4 bugs)
+- Parquet I/O errors (1 bug)
+
+## Key Takeaways
+
+Analyzing the uncovered bugs led to some key learnings to avoid common pitfalls:
+
+1. Thoroughly validate function parameters and edge cases. Main cases include:
+   - For `String` parameters, check for empty strings.
+   - For `Int` parameters, check for large number values.
+   - When supporting any parameter type, consider unusual types like `Null`, `EmptyArray`, `Bitmap` etc.
+   - When only supporting specific types, see if other types can auto-convert or check and return errors early.
+2. Use unwrap judiciously. Explicitly handle `Result` and `Option` instead of assuming success.
+3. Understand SQL semantic rules during development e.g. constraints around `GROUP BY`, `ORDER BY` etc. Perform semantic checks in the Binder phase to prevent runtime errors.
+4. Add more unit test cases for critical modules to prevent bugs from internal logic errors.
+
+## What's Next
+
+SQLsmith generates most common SQLs and integrates with CI. Upcoming improvements include:
```

Comment:
SQLsmith helps enhance Databend's stability and reliability by uncovering hidden issues. It generates common SQL queries, integrates with CI, and upcoming improvements include: